doc: describe why new Buffer() is problematic (#28825)
nodejs/node: doc/api/buffer.md

differently based on what arguments are provided:

* Passing a number as the first argument to `Buffer()` (e.g. `new Buffer(10)`)

>   allocates a new `Buffer` object of the specified size. Prior to Node.js v8.0.0,

Reviewer: In prose, we've dropped the `v` to avoid confusion with V8. Phonetically, it's just "node 8" and not "node v8" and our writing should reflect that. We'll never get it out of the version string returned by `--version` and `-v`, but we do what we can.

```suggestion
  allocates a new `Buffer` object of the specified size. Prior to Node.js 8.0.0,
```

  the memory allocated for such `Buffer` instances is *not* initialized and
  *can contain sensitive data*. Such `Buffer` instances *must* be subsequently
  initialized by using either [`buf.fill(0)`][`buf.fill()`] or by writing to the
  entire `Buffer`. While this behavior is *intentional* to improve performance,
  development experience has demonstrated that a more explicit distinction is
  required between creating a fast-but-uninitialized `Buffer` versus creating a

>   slower-but-safer `Buffer`. Since Node.js v8.0.0, `Buffer(num)` and `new

Reviewer: ```suggestion
  slower-but-safer `Buffer`. Since Node.js 8.0.0, `Buffer(num)` and `new
```

into applications when argument validation or `Buffer` initialization is not
performed.

> For example, if an application can be convinced to call `new Buffer(100)`

Reviewer: ```suggestion
For example, if an application calls `new Buffer(100)`
```

Author: the convincing to do this instead of the other is what is important here, I think. Maybe some other wording would make that more clear?

Reviewer: Maybe this?

```suggestion
For example, if an attacker can cause an application to call `new Buffer(100)`
```

...or maybe this?

```suggestion
For example, if a bug makes it possible for an application to call `new Buffer(100)`
```

Author: How about "if an attacker can cause a number to be passed to an application that assumes it will always be passed strings, it may call `new .."?

Reviewer: That has the virtue of clarity. 👍 That's good enough for me. I'd use somewhat different wording, but either yours above or mine below is fine by me:

If an attacker can cause an application to receive a number where a string is expected, the application may call `new....

Author: used yours

instead of `new Buffer("100")`, it will allocate a 100 byte buffer instead
of allocating a 3 byte buffer with content `"100"`. This is commonly possible

> using JSON API calls, since JSON distinguishes between numeric and string types,

Reviewer: ```suggestion
using JSON API calls. Since JSON distinguishes between numeric and string types,
```

it allows injection of numbers where a naive application might expect to always

> receive a string.  Before Node.js v8.0.0, the 100 byte buffer might contain

Reviewer: ```suggestion
receive a string.  Before Node.js 8.0.0, the 100 byte buffer might contain
```

arbitrary pre-existing in-memory data, so may be used to expose in-memory

> secrets to a remote attacker.  Since Node.js v8.0.0, exposure of memory can

Reviewer: ```suggestion
secrets to a remote attacker.  Since Node.js 8.0.0, exposure of memory can
```

> *not* occur, the data is zero-filled. However, other attacks are still

Reviewer: ```suggestion
not occur because the data is zero-filled. However, other attacks are still
```

> not occur because the data is zero-filled. However, other attacks are still

Reviewer 2: I wonder if it's really an advantage to include the description about old Node.js versions. Since the documentation is meant for newer versions, it's probably best to limit the description to the remaining attack vectors?

Author: If you expand the context diff upwards, you will see that this information is in an entire section dedicated to describing the situation in 6.x: how it used to work, why it was a problem, and guidelines on how to update from code developed pre-8.x to the new APIs.

  initialized `Buffer` of the specified size. This method is slower than
  [`Buffer.allocUnsafe(size)`][`Buffer.allocUnsafe()`] but guarantees that newly
  created `Buffer` instances never contain old data that is potentially

>   sensitive. A `TypeError` will be thrown if size is not a number.

Reviewer: ```suggestion
  sensitive. A `TypeError` will be thrown if `size` is not a number.
```

* [`Buffer.allocUnsafe(size)`][`Buffer.allocUnsafe()`] and
  [`Buffer.allocUnsafeSlow(size)`][`Buffer.allocUnsafeSlow()`] each return a
  new uninitialized `Buffer` of the specified `size`. Because the `Buffer` is
  uninitialized, the allocated segment of memory might contain old data that is

>   potentially sensitive. A `TypeError` will be thrown if size is not a number.

Reviewer: ```suggestion
  potentially sensitive. A `TypeError` will be thrown if `size` is not a number.
```

>   potentially sensitive. A `TypeError` will be thrown if `size` is not a number.

Reviewer 2: We rarely document any error cases and I personally would rather keep them out of the docs to keep the description brief.

Author: error cases are documented in buffer.md:
```
% grep "will be thrown" doc/api/buffer.md 
  sensitive. A `TypeError` will be thrown if `size` is not a number.
A `TypeError` will be thrown if `size` is not a number.
A `TypeError` will be thrown if `size` is not a number.
A `TypeError` will be thrown if `size` is not a number.
A `TypeError` will be thrown if `array` is not an `Array` or other type
A `TypeError` will be thrown if `arrayBuffer` is not an [`ArrayBuffer`][] or a
A `TypeError` will be thrown if `buffer` is not a `Buffer` or other type
A `TypeError` will be thrown if `object` has not mentioned methods or is not of
A `TypeError` will be thrown if `string` is not a string or other type
```

But given its documented elsewhere, I removed it from here.

Node.js can be started using the `--zero-fill-buffers` command line option to
cause all newly allocated `Buffer` instances to be zero-filled upon creation by

> default. Before Node.js v8.0.0 this included buffers allocated by `new

Reviewer: ```suggestion
default. Before Node.js 8.0.0, this included buffers allocated by `new
```

> Buffer(size)`, but since v8.0.0 buffers allocated with `new` are always

Reviewer: ```suggestion
Buffer(size)`. Since Node.js 8.0.0, buffers allocated with `new` are always
```

[`buffer.constants.MAX_LENGTH`][] or smaller than 0, [`ERR_INVALID_OPT_VALUE`][]
is thrown. A zero-length `Buffer` is created if `size` is 0.

> Prior to Node.js v8.0.0, the underlying memory for `Buffer` instances

Reviewer: ```suggestion
Prior to Node.js 8.0.0, the underlying memory for `Buffer` instances
```
